source/opt/spread_volatile_semantics.cpp

```diff
@@ -0,0 +1,143 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/spread_volatile_semantics.h"
+
+#include "source/opt/decoration_manager.h"
+#include "source/opt/ir_builder.h"
+
+namespace spvtools {
+namespace opt {
+namespace {
+
+const uint32_t kOpDecorateInOperandBuiltinDecoration = 2u;
+const uint32_t kOpLoadInOperandMemoryOperands = 1u;
+
+bool IsBuiltInForVolatileSemantics(uint32_t built_in) {
+  switch (built_in) {
+    case SpvBuiltInSMIDNV:
+    case SpvBuiltInWarpIDNV:
+    case SpvBuiltInSubgroupSize:
+    case SpvBuiltInSubgroupLocalInvocationId:
+    case SpvBuiltInSubgroupEqMask:
+    case SpvBuiltInSubgroupGeMask:
+    case SpvBuiltInSubgroupGtMask:
+    case SpvBuiltInSubgroupLeMask:
+    case SpvBuiltInSubgroupLtMask:
+      return true;
+    default:
+      return false;
+  }
+}
+
+bool HasVolatileDecoration(analysis::DecorationManager* decoration_manager,
+                           uint32_t var_id) {
+  return decoration_manager->HasDecoration(var_id, SpvDecorationVolatile);
+}
+
+}  // namespace
+
+Pass::Status SpreadVolatileSemantics::Process() {
+  Status status = Status::SuccessWithoutChange;
+  bool is_vk_memory_model_enabled = context()->get_feature_mgr()->HasCapability(
+      SpvCapabilityVulkanMemoryModel);
+  for (Instruction& var : context()->types_values()) {
+    if (!IsTargetForVolatileSemantics(&var)) {
+      continue;
+    }
+
+    if (is_vk_memory_model_enabled) {
+      SetVolatileForLoads(&var);
+    } else {
+      DecorateVarWithVolatile(&var);
+    }
+    status = Status::SuccessWithChange;
+  }
+  return status;
+}
+
+void SpreadVolatileSemantics::DecorateVarWithVolatile(Instruction* var) {
+  analysis::DecorationManager* decoration_manager =
+      context()->get_decoration_mgr();
+  uint32_t var_id = var->result_id();
+  if (HasVolatileDecoration(decoration_manager, var_id)) {
+    return;
+  }
+  get_decoration_mgr()->AddDecoration(
+      SpvOpDecorate,
+      {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {var_id}},
+       {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {SpvDecorationVolatile}}});
+}
+
+void SpreadVolatileSemantics::SetVolatileForLoads(Instruction* var) {
+  std::vector<Instruction*> updated_insts;
+  auto* def_use_mgr = context()->get_def_use_mgr();
+  def_use_mgr->ForEachUser(var, [&updated_insts](Instruction* user) {
+    if (user->opcode() != SpvOpLoad) {
+      return;
+    }
+
+    updated_insts.push_back(user);
+    if (user->NumInOperands() <= kOpLoadInOperandMemoryOperands) {
+      user->AddOperand({SPV_OPERAND_TYPE_ID, {SpvMemoryAccessVolatileMask}});
+      return;
+    }
+
+    uint32_t memory_operands =
+        user->GetSingleWordInOperand(kOpLoadInOperandMemoryOperands);
+    memory_operands |= SpvMemoryAccessVolatileMask;
+    user->SetInOperand(kOpLoadInOperandMemoryOperands, {memory_operands});
+  });
+
+  for (auto* inst : updated_insts) {
```

Comment:
Is this necessary? The only change made is a non-id based operand. No def-use chains will change.